### PR TITLE
Decoding Functions

### PR DESCRIPTION
**Pull request by Bridgerz**

### CODE REVIEW

File: contracts/utils/BridgeMessage.sol [hunk 1]
```diff
@@ -197,28 +197,61 @@ library BridgeMessage {
...
+        for (uint8 i = 0; i < membersLength; i++) {
+            // Calculate the starting index for each address
+            offset += i * 20;
+            address member;
+            // Extract each address
+            assembly {
+                member := mload(add(add(payload, 20), offset))
+            }
+            // Store the extracted address
+            members[i] = member;
```
Comment (longbowlu) [inline contracts/utils/BridgeMessage.sol:214]: what happens when length is more than what is in payload? should we check payload.length() matches?

File: contracts/utils/BridgeMessage.sol [hunk 2]
```diff
@@ -197,28 +197,61 @@ library BridgeMessage {
...
+        }
+        // blocklistType: 0 = blocklist, 1 = unblocklist
+        bool blocklisted = (blocklistType == 0) ? true : false;
```
Comment (longbowlu) [inline contracts/utils/BridgeMessage.sol]:
```suggestion
        bool blocklisted = blocklistType == 0;
```

does this work?

File: contracts/utils/BridgeMessage.sol [hunk 3]
```diff
@@ -197,28 +197,61 @@ library BridgeMessage {
...
+        return (blocklisted, members);
     }
 
     function decodeEmergencyOpPayload(bytes memory payload) internal pure returns (bool) {
-        (uint8 emergencyOpCode) = abi.decode(payload, (uint8));
+        require(payload.length == 1, "BridgeMessage: Invalid payload length");
+        uint8 emergencyOpCode = uint8(payload[0]);
         require(emergencyOpCode <= 1, "BridgeMessage: Invalid op code");
         return emergencyOpCode == 0 ? true : false;
```
Comment (longbowlu) [inline contracts/utils/BridgeMessage.sol]:
```suggestion
        return emergencyOpCode == 0;
```

File: contracts/utils/BridgeMessage.sol [hunk 4]
```diff
@@ -197,28 +197,61 @@ library BridgeMessage {
...
 
-    function decodeUpdateLimitPayload(bytes memory payload) internal pure returns (uint256) {
-        (uint256 newLimit) = abi.decode(payload, (uint256));
-        return newLimit;
+    function decodeUpdateLimitPayload(bytes memory payload)
+        internal
+        pure
+        returns (uint8 senderChainID, uint64 newLimit)
+    {
+        require(payload.length >= 9, "BridgeMessage: Invalid payload length");
```
Comment (longbowlu) [inline contracts/utils/BridgeMessage.sol]: should be `== 9`?

File: contracts/utils/BridgeMessage.sol [hunk 5]
```diff
@@ -197,28 +197,61 @@ library BridgeMessage {
...
 
     function decodeUpdateAssetPayload(bytes memory payload)
         internal
         pure
-        returns (uint8, uint256)
+        returns (uint8 tokenID, uint64 assetPrice)
     {
-        (uint8 tokenId, uint256 price) = abi.decode(payload, (uint8, uint256));
-        return (tokenId, price);
+        require(payload.length >= 9, "BridgeMessage: Invalid payload length");
```
Comment (longbowlu) [inline contracts/utils/BridgeMessage.sol]: ditto

File: contracts/utils/BridgeMessage.sol [hunk 6]
```diff
@@ -197,28 +197,61 @@ library BridgeMessage {
...
+    function decodeUpdateLimitPayload(bytes memory payload)
+        internal
+        pure
+        returns (uint8 senderChainID, uint64 newLimit)
+    {
+        require(payload.length >= 9, "BridgeMessage: Invalid payload length");
+        senderChainID = uint8(payload[0]);
+
+        // Extracts the uint64 value by loading 32 bytes starting just after the first byte.
+        // Position uint64 to the least significant bits by shifting it 192 bits to the right.
```
Comment (longbowlu) [inline contracts/utils/BridgeMessage.sol:237]: love the comments!

File: test/BridgeLimiterTest.t.sol [hunk 7]
```diff
@@ -149,6 +151,140 @@ contract BridgeLimiterTest is BridgeBaseTest {
         // Call the updateLimitWithSignatures function
         limiter.updateLimitWithSignatures(signatures, message);
 
-        assertEq(limiter.totalLimit(), 100000000);
+        assertEq(limiter.totalLimit(), 1000000000);
```
Comment (longbowlu) [inline test/BridgeLimiterTest.t.sol:156]: nit: assertEq old limit in the beginning?

File: test/CommitteeUpgradeableTest.t.sol [hunk 8]
```diff
@@ -182,5 +183,149 @@ contract CommitteeUpgradeableTest is BridgeBaseTest {
...
+
+        bytes memory payload = abi.encodePacked(
+            abi.encode(address(bridge)), abi.encode(address(bridgeV2)), emptyCallData
+        );
+
+        bytes memory encodedMessage = abi.encodePacked(messagePrefix, payload);
+
+        bytes32 messageHash = keccak256(encodedMessage);
+
+        // // Create transfer message
```
Comment (longbowlu) [inline test/CommitteeUpgradeableTest.t.sol]:
```suggestion
        // Create transfer message
```

File: test/CommitteeUpgradeableTest.t.sol [hunk 9]
```diff
@@ -182,5 +183,149 @@ contract CommitteeUpgradeableTest is BridgeBaseTest {
...
+
+        bytes memory payload = abi.encodePacked(
+            abi.encode(address(bridge)), abi.encode(address(bridgeV2)), newMockFunc2CallData
+        );
+
+        bytes memory encodedMessage = abi.encodePacked(messagePrefix, payload);
+
+        bytes32 messageHash = keccak256(encodedMessage);
+
+        // // Create transfer message
```
Comment (longbowlu) [inline test/CommitteeUpgradeableTest.t.sol]:
```suggestion
        // Create transfer message
```

File: test/CommitteeUpgradeableTest.t.sol [hunk 10]
```diff
@@ -182,5 +183,149 @@ contract CommitteeUpgradeableTest is BridgeBaseTest {
...
+
+        bytes memory payload = abi.encodePacked(
+            abi.encode(address(bridge)), abi.encode(address(bridgeV2)), newMockFunc1CallData
+        );
+
+        bytes memory encodedMessage = abi.encodePacked(messagePrefix, payload);
+
+        bytes32 messageHash = keccak256(encodedMessage);
+
+        // // Create transfer message
```
Comment (longbowlu) [inline test/CommitteeUpgradeableTest.t.sol]:
```suggestion
        // Create transfer message
```

File: test/CommitteeUpgradeableTest.t.sol [hunk 11]
```diff
@@ -182,5 +183,149 @@ contract CommitteeUpgradeableTest is BridgeBaseTest {
...
+
+        bytes memory payload = abi.encodePacked(
+            abi.encode(address(bridge)), abi.encode(address(bridgeV2)), initV2CallData
+        );
+
+        bytes memory encodedMessage = abi.encodePacked(messagePrefix, payload);
+
+        bytes32 messageHash = keccak256(encodedMessage);
+
+        // // Create transfer message
```
Comment (longbowlu) [inline test/CommitteeUpgradeableTest.t.sol]:
```suggestion
        // Create transfer message
```

File: contracts/utils/BridgeMessage.sol (hunk 2 above)
Comment (Bridgerz) [reply contracts/utils/BridgeMessage.sol]: yes, lol. This must be co-pilot generated.

File: contracts/utils/BridgeMessage.sol (hunk 1 above)
Comment (Bridgerz) [reply contracts/utils/BridgeMessage.sol:214]: The function will cause a revert. We could do: 

`require((payload.length - offset) % 20 == 0, "BridgeMessage: Invalid payload length");`

File: contracts/SuiBridge.sol [hunk 12]
```diff
@@ -39,14 +48,20 @@ contract SuiBridge is ISuiBridge, CommitteeUpgradeable, PausableUpgradeable {
...
     ) external initializer {
         __CommitteeUpgradeable_init(_committee);
         __Pausable_init();
         tokens = IBridgeTokens(_tokens);
         vault = IBridgeVault(_vault);
         limiter = IBridgeLimiter(_limiter);
         weth9 = IWETH9(_weth9);
+
+        for (uint8 i = 0; i < _supportedChainIDs.length; i++) {
+            // require(_supportedChainIDs[i] != BridgeCommittee.chainID, "SuiBridge: Cannot support self");
```
Comment (longbowlu) [inline contracts/SuiBridge.sol:55]: uncomment this?

File: contracts/SuiBridge.sol [hunk 13]
```diff
@@ -61,18 +76,15 @@ contract SuiBridge is ISuiBridge, CommitteeUpgradeable, PausableUpgradeable {
...
 
-        // TODO:
-        // require(supportedChainIDs[message.chainID]);
-
         BridgeMessage.TokenTransferPayload memory tokenTransferPayload =
             BridgeMessage.decodeTokenTransferPayload(message.payload);
 
-        // TODO:
-        // require(tokenTransferPayload.targetChain == committee.chainID());
+        // require(tokenTransferPayload.targetChain == BridgeCommittee.chainID());
```
Comment (longbowlu) [inline contracts/SuiBridge.sol]: uncomment?

**Review by longbowlu — COMMENTED** [review]
for supporteChains, can we add test cases where unsupported chain is rejected for both sending and receiving?

**Comment by longbowlu** [comment]
oh I missed your final commit, looking

**Review by longbowlu — APPROVED** [review]

**Comment by longbowlu** [comment]
awesome work!

**Pull request merged**